kong/plugins/prometheus/wasmx.lua

```diff
@@ -0,0 +1,234 @@
+  end
+
+  labels_serialization_buf:skip(1)  -- discard leading comma
+
+  return "{" .. labels_serialization_buf:get() .. "}"
+end
+
+
+local function serialize_metric(m, buf)
+  buf:putf("# HELP %s\n# TYPE %s %s", m.name, m.name, m.type)
+
+  if m.type == "histogram" then
+    sum_lines_buf:reset()
+    count_lines_buf:reset()
+
+    for _, pair in ipairs(m.labels) do
+      local count, sum = 0, 0
+      local labels, labeled_m = pair[1], pair[2]
+      local slabels, blabels = "", "{"
+
+      if #labels > 0 then
+        slabels = serialize_labels(labels)
+        blabels = slabels:sub(1, #slabels - 1) .. ","
+      end
+
+      for _, bin in ipairs(labeled_m.value) do
+        count = count + bin.count
+
+        buf:putf('\n%s%sle="%s"} %s',
+                 m.name,
+                 blabels,
+                 (bin.ub ~= 4294967295 and bin.ub or "+Inf"),
+                 count)
+      end
+
+      sum = sum + labeled_m.sum
+
+      sum_lines_buf:putf("\n%s_sum%s %s", m.name, slabels, sum)
+      count_lines_buf:putf("\n%s_count%s %s", m.name, slabels, count)
+    end
+
+    buf:put(sum_lines_buf:get())
+    buf:put(count_lines_buf:get())
+
+  else
+    assert(m.type == "gauge" or m.type == "counter", "unknown metric type")
+
+    for _, pair in ipairs(m.labels) do
+      local labels, labeled_m = pair[1], pair[2]
+      local slabels = (#labels > 0) and serialize_labels(labels) or ""
+
+      buf:putf("\n%s%s %s", m.name, slabels, labeled_m.value)
+    end
+  end
+
+  buf:put("\n")
+end
+
+
+local function require_wasmx()
+  if not wasmx_shm then
+    local ok, _wasmx_shm = pcall(require, "resty.wasmx.shm")
+    if ok then
+      wasmx_shm = _wasmx_shm
+    end
+  end
+end
+
+
+_M.metrics_data = function()
+  if not wasm.enabled() then
+    return
+  end
+
+  local metrics = {}
+  local parsed = {}
+
+  -- delayed require of the WasmX module, to ensure it is loaded
+  -- after ngx_wasm_module.so is loaded.
+  require_wasmx()
+
+  if not wasmx_shm then
+    return
+  end
+
+  wasmx_shm.metrics:lock()
+
+  for key in wasmx_shm.metrics:iterate_keys() do
+    local pair = { key, wasmx_shm.metrics:get_by_name(key, GET_METRIC_OPTS) }
+    table_insert(metrics, pair)
+  end
+
+  wasmx_shm.metrics:unlock()
+
+  -- in WasmX the different labels of a metric are stored as separate metrics;
+  -- aggregate those separate metrics into a single one.
+  for _, pair in ipairs(metrics) do
+    local key = pair[1]
+    local m = pair[2]
+    local name, labels = parse_key(key)
+
+    parsed[name] = parsed[name] or { name = name, type = m.type, labels = {} }
+
+    table_insert(parsed[name].labels, { labels, m })
+  end
+
+  metrics_data_buf:reset()
+
+  for i, metric_by_label in sorted_pairs(parsed) do
+    metrics_data_buf:put(serialize_metric(metric_by_label, metrics_data_buf))
+
+    if i % FLUSH_EVERY == 0 then
+      ngx_say(metrics_data_buf:get())
+    end
+  end
+
+  ngx_say(metrics_data_buf:get())
+end
+
+
+return _M
```

Comment:
Not a blocker (the PR is in good shape, and I've held it up for long enough already), but a note for future us if we amend this file:

The integration tests you've added provide good coverage for this feature, but I think it would be useful to have unit tests for the private `serialize_*` and `parse_*` functions in this module, both to serve as living documentation of their behavior and to make it easier to fix any bugs that we may find later on.